Employ bouncer to perform operation authorization

keynslug opened the pull request:
### TODO

* [x] Proper `capi_bouncer` interface and context builders.
* [x] Simplify / enrich bouncer-proto w/ additional contexts.
* [ ] Handle restrictions where applicable.

Commit:
Bump deps w/ swithing to public enpoints

* rbkmoney/bender-proto@0d5813b8
* rbkmoney/cowboy_access_log@c058ad42
* rbkmoney/cowboy_cors@5a3b084f
* rbkmoney/logger_logstash_formatter@498fda5f

Commit:
Add rbkmoney/bouncer_client_erlang@d053efc9 dep

Commit:
Drop dead code around `service_type`

Commit:
Switch deps to public endpoints

Commit:
Rework authorizer module

Add key metadata to aid bouncer integration

Commit:
[wip] Rework operation authorization

* Introduce providers to capi_auth: bouncer + legacy.
* Perform authz inside each operation handler, since there's now too
   much specifics to do it once for all operations.

Commit:
Drop api key blacklisting

This is bouncer responsibility now. We still need to pass bouncer
something to identify individual tokens through.

Commit:
Merge remote-tracking branch 'origin/master' into MSPF-603/ft/bouncer-auth

Commit:
Fix borked multiline edits

Commit:
Integrate rest of operations w/ new auth facility

Commit:
Refine types and assertions

Commit:
Ensure jwt verify returns key metadata

Commit:
Add accessor for subject email claim

Commit:
Refine auth function names for clarity a bit

Commit:
Set auth failure reason to 'policy' when asking bouncer

Commit:
Log party id when creating it implicitly

Commit:
Add capi-specific context builders

Adapt handler and auth code to these new abstractions.

Commit:
Drop blacklist instantiation also

We still need to solve token blacklisting though, for this we will need to
pass some kind of token identifier to bouncer.

Commit:
Upgrade to rbkmoney/woody_erlang@58f56b4

Also upgrade dmt_client since it also uses the same woody client code.

Commit:
Avoid extraneous context wrapping

Commit:
Bump to rbkmoney/bouncer_client_erlang@73744df

Commit:
Downgrade FIXME into a simple TODO

WWWcool commented on apps/capi/src/capi_bouncer.erl:
```diff
@@ -0,0 +1,164 @@
+        #{
+            id => UserID,
+            email => capi_authorizer_jwt:get_subject_email(Claims),
+            realm => #{id => maps:get(user_realm, Metadata, undefined)}
```
а realm undefined - это не ошибка? тут либо валидное дефолтное значение либо ломаться?

WWWcool commented on apps/capi/src/capi_auth.erl:
```diff
@@ -127,7 +190,18 @@ issue_invoice_template_access_token(PartyID, InvoiceTplID, Claims) ->
+            scope => [
+                #{party => #{id => PartyID}}
+                % FIXME
+                % #{invoice_template => #{id => InvoiceTplID}}
```
забытое?

keynslug replied:
Поправил.

WWWcool commented on apps/capi/src/capi_auth.erl:
```diff
@@ -143,57 +217,83 @@ issue_customer_access_token(PartyID, CustomerID, Claims) ->
+            expiration => make_auth_expiration(ExpiresAt),
+            scope => [
+                #{party => #{id => PartyID}}
+                % FIXME
```
и тут?

keynslug replied:
Поправил.

Commit:
Do not imply operation ctx might not be present

Commit:
Bump to rbkmoney/bouncer_client_erlang@921f52f

WWWcool commented on apps/capi/src/capi_bouncer_context.erl:
```diff
@@ -0,0 +1,340 @@
+-export_type([acc/0]).
+-export_type([fragments/0]).
+
+-type prototype() :: [
```
предложил бы сделать отдельно prototype и добавить к нему prototypes или prototype_list, текущая ситуация усложняет чтение когда, не сразу понятно что везде где упоминается прототип в единственном числе на самом деле список

WWWcool commented on apps/capi/src/capi_authorizer_jwt.erl:
```diff
@@ -5,36 +5,51 @@
+-type realm() :: binary().
+-type expiration() :: non_neg_integer() | unlimited.
+-type auth_method() ::
+    user_session_token.
```
а на этом уровне нам не нужно вводить другие типы токенов?

keynslug replied:
В этом PR не видится в этом необходимости. Потом быть может `api_key_token` какой-нибудь появится.

Commit:
Add bouncer_client config example

WWWcool commented on apps/capi/src/capi_real_handler.erl:
```diff
@@ -448,12 +553,23 @@ process_request('SearchInvoices', Req, Context, ReqCtx) ->
+    InvoiceID = genlib_map:get('invoiceID', Req),
+    OpCtx = with_maybe(invoice, InvoiceID, with_maybe(shop, ShopID, #{party => PartyID})),
+    % TODO
+    % No payment ctx here, need to add it or simplify proto.
```
даа, придется видимо политики ослабить для методов статистики

WWWcool commented on apps/capi/src/capi_real_handler.erl:
```diff
@@ -1070,143 +1333,181 @@ process_request('GetPaymentInstitutions', Req, _Context, ReqCtx) ->
+        % This is not entirely correct, yet sufficient for a legacy API.
+        {exception, #payproc_InvalidUser{}} ->
+            reply(200, [], ReqSt1);
+        {exception, #payproc_PartyNotFound{}} ->
```
кажется когда с Юрой добавляли отсутствие пати, то решили ломаться если его нет, а не прокидывать ошибку

keynslug replied:
Я исходил из предпосылки, что абстрактный пользователь всех наших разномастных api действительно может дёрнуть что-нибудь в swag-v1, не создав до этого, например, организацию. То есть кажется обрабатывать надо в штатном режиме эту ситуацию, отдавая клиентскую ошибку. Или ты про то, что мы теперь не увидим поток неуспешных запросов, отбившихся по причине отсутствия party?

WWWcool replied:
ну мы в прошлый раз дошли уже до свага и начали эту ошибку в него добавлять)) потом сходили к Андрею, он сказал, что пользователь все равно в большей части случае ничего не сможет сделать с такой ошибкой, поэтому нет смысла ее отдавать и лучше сразу ломаться - это быстрее до нас дойдет

keynslug replied:
> ничего не сможет сделать с такой ошибкой, 

Но ведь он может [создать организацию](https://github.com/rbkmoney/swag-organizations/blob/9e36e8e1e4b7ba8b04db1e94a3908ab547c19283/openapi/paths/orgs.yaml#L3), ну разве что orgmgmt нужен работоспособный.

С другой стороны трогать легаси спеку и добавлять туда новые клиентские ошибки совершенно не хочется.

Commit:
Fix sys.config example

Commit:
Update to rbkmoney/thrift_erlang@846a081

Commit:
Fix JWT claim verification

Commit:
Expect 'email' claim might be absent

Commit:
Fix payment / refund accessors

Commit:
Mention 'graceful_shutdown_timeout' in sys.config

Commit:
Drop blacklist leftovers from sys.config

Commit:
Fix partymgmt function name

Commit:
Implement compute terms a bit dumber

Commit:
Restore "0 is unlimited" exp claim semantics

Commit:
Partially adopt proper formatting

Commit:
Drop blacklisting-related testcase

Commit:
Use woody to choose free port for mock servers

Commit:
Use separate keys in testcases w/ legacy tokens

Commit:
Make sure capi stops right away in tests

Commit:
Toss cases between groups to reflect requirements better

Commit:
Mock PullRange method also for recent dmt_client

Commit:
Provide sufficient mocks in tests for bouncer facilities

Also make sure to update them for recent woody.

Commit:
Fix matching on authctx metadata

Commit:
Rename `prototype()` to `prototypes()`

Commit:
Make rebar3_fmt a bit happier

Commit:
Fix operation access clause

Commit:
Reframe specific access token tests for legacy tokens

Commit:
Tell dialyzer to ignore mistypings of foolproof functions

Sadly couldn't come up with a prettier solution.

Commit:
Fix some last bits of formatting

Commit:
Supply 'jti' as part of bouncer ctx

* Bump to rbkmoney/bouncer_client_erlang@f01fc16

So we can solve token blocklisting problem on the bouncer-policies side.

Commit:
Stop sending extraneous bouncer context

Only w/ invoices / templates / customers for the time being.
Some extra context still there to get rid of.

Commit:
Fix couple misplaced comments

Commit:
Fix `DownloadFile` bouncer context

Also drop ownership control logic.

Commit:
Actualize search ops bouncer contexts

Commit:
Fix formatting in tests

Commit:
Fix `DownloadFile` testcase

Verify that incomplete report is not "visible" instead.

Commit:
Simplify `CreateWebhook` bouncer context

It's ugly and we'll deal with it later.

Commit:
Fix few mistypings

Commit:
Upgrade to rbkmoney/bouncer-ptoto@290b5b1

Commit:
Switch to rbkmoney/bouncer-proto@a126059 @ master

Commit:
Verify expected ctx w/ bouncer mocks in tests

Commit:
Bump to rbkmoney/bouncer_client_erlang@36cb53a

Commit:
Fix formatting quirk

WWWcool commented on apps/capi/src/capi_bouncer.erl:
```diff
@@ -0,0 +1,159 @@
+%%
+
+extract_context_fragments_by(claim, {Claims, _}, _) ->
+    % FIXME handle errors
```
тут видимо выход на некую миграцию пока не существующую будет?

keynslug replied:
Кстати. Да, это вполне вероятно, переделаю на TODO. Шансов словить в этом месте `{error, {unsupported, ...}}` как будто бы нет, не вижу вариантов, когда мы из верифицированного токена кривые данные можем достать в процессе штатной работы.

keynslug replied:
Кстати. Да, это вполне вероятно, переделаю на TODO. Шансов словить в этом месте `{error, {unsupported, ...}}` как будто бы нет, не вижу вариантов, когда мы из верифицированного токена кривые данные можем достать в процессе штатной работы.

WWWcool commented on apps/capi/test/capi_tests_SUITE.erl:
```diff
@@ -303,115 +311,68 @@ groups() ->
+    [{context, get_context(Token)}, {group_apps, Apps} | Config];
+init_per_group(operations_by_base_api_token, Config) ->
+    SupPid = start_mocked_service_sup(),
+    % Apps1 = mock_bouncer_client(judge_always_forbidden(), SupPid),
```
забытое?

WWWcool commented on apps/capi/test/capi_tests_SUITE.erl:
```diff
@@ -523,33 +505,31 @@ authorization_error_no_header_test(_Config) ->
         ],
         Config
     ),
+    mock_bouncer_client(
```
кажется в большинстве случаев эти моки можно разделить на несколько больших групп - инвойсы, пати, шалоны и тд и еще больше сократить эту запись

keynslug replied:
Хорошо бы пример небольшой, а то не до конца понятно, как ты это видишь.

WWWcool replied:
```
mock_bouncer_client(binding, OpID, Config) ->
    mock_bouncer_client(
        ?assertContextMatches(
            #bctx_v1_ContextFragment{
                capi = ?CTX_CAPI(?CTX_BINDING_OP(OpID, ?STRING, ?STRING)),
                payment_processing = #bctx_v1_ContextPaymentProcessing{
                    customer = ?CTX_CUSTOMER(?STRING, ?STRING, ?STRING)
                }
            }
        ),
        Config
    );
mock_bouncer_client(customer, OpID, Config) ->
    mock_bouncer_client(
        ?assertContextMatches(
            #bctx_v1_ContextFragment{
                capi = ?CTX_CAPI(?CTX_CUSTOMER_OP(OpID, ?STRING)),
                payment_processing = #bctx_v1_ContextPaymentProcessing{
                    customer = ?CTX_CUSTOMER(?STRING, ?STRING, ?STRING)
                }
            }
        ),
        Config
    ).
```

Commit:
Drop commented out code for good

Commit:
Also handle malformed base64 binaries in claims

Commit:
Outline why no error handling in claim decoding

Commit:
Remove bits of redundant code

Commit:
Pass token ID also for custom access tokens

Commit:
Remove unused sup from one of test groups

Commit:
Factor out most of bouncer mocks boilerplate

Commit:
Fix formatting

Commit:
Rework injecting token id for bouncer-client compat

Commit:
Also verify expected bouncer ctx by token class

Commit:
Fix redundant multiple scopes in access token

Commit:
Fix formatting
